roofit/histfactory/src/ParamHistFunc.cxx

```diff
@@ -521,83 +511,71 @@ RooArgList ParamHistFunc::createParamSet(const std::string& Prefix, Int_t numBin
+    return _binMapVector;
   }
 
-  Int_t numBinsX = 1;
-  Int_t numBinsY = 1;
-  Int_t numBinsZ = 1;
+  int numVars = _dataVars.size();
+
+  int numBinsX = 1;
+  int numBinsY = 1;
+  int numBinsZ = 1;
 
   if( numVars == 1 ) {
-    RooRealVar* varX = (RooRealVar*) _dataVars.at(0);
-    numBinsX = varX->numBins();
-    numBinsY = 1;
-    numBinsZ = 1;
+    numBinsX = static_cast<RooRealVar const&>(_dataVars[0]).numBins();
   } else  if( numVars == 2 ) {
-    RooRealVar* varX = (RooRealVar*) _dataVars.at(0);
-    RooRealVar* varY = (RooRealVar*) _dataVars.at(1);
-    numBinsX = varX->numBins();
-    numBinsY = varY->numBins();
-    numBinsZ = 1;
+    numBinsX = static_cast<RooRealVar const&>(_dataVars[0]).numBins();
+    numBinsY = static_cast<RooRealVar const&>(_dataVars[1]).numBins();
   } else  if( numVars == 3 ) {
-    RooRealVar* varX = (RooRealVar*) _dataVars.at(0);
-    RooRealVar* varY = (RooRealVar*) _dataVars.at(1);
-    RooRealVar* varZ = (RooRealVar*) _dataVars.at(2);
-    numBinsX = varX->numBins();
-    numBinsY = varY->numBins();
-    numBinsZ = varZ->numBins();
+    numBinsX = static_cast<RooRealVar const&>(_dataVars[0]).numBins();
+    numBinsY = static_cast<RooRealVar const&>(_dataVars[1]).numBins();
+    numBinsZ = static_cast<RooRealVar const&>(_dataVars[2]).numBins();
   } else {
     std::cout << "ParamHistFunc() - Only works for 1-3 variables (1d-3d)" << std::endl;
     throw -1;  
   }
 
-  // Fill the mapping between
-  // RooDataHist bins and TH1 Bins:
+  // Fill the map
+  _binMapVector.resize(numBinsX * numBinsY * numBinsZ);
+  for( int i = 0; i < numBinsX; ++i ) {
+    for( int j = 0; j < numBinsY; ++j ) {
+      for( int k = 0; k < numBinsZ; ++k ) {
 
-  // Clear the map
-  _binMap.clear();
+        int rooDataSetBin = k + j*numBinsZ + i*numBinsY*numBinsZ; 
+        int histBin    = i + j*numBinsX + k*numBinsX*numBinsY; 
 
-  // Fill the map
-  for( Int_t i = 0; i < numBinsX; ++i ) {
-    for( Int_t j = 0; j < numBinsY; ++j ) {
-      for( Int_t k = 0; k < numBinsZ; ++k ) {
-	
-	Int_t RooDataSetBin = k + j*numBinsZ + i*numBinsY*numBinsZ; 
-	Int_t TH1HistBin    = i + j*numBinsX + k*numBinsX*numBinsY; 
-	  
-	_binMap[RooDataSetBin] = TH1HistBin;
-	
+        _binMapVector[rooDataSetBin] = histBin;
+        
       }
     }
   }
   
-  return 0;
+  return _binMapVector;
+}
+
 
+////////////////////////////////////////////////////////////////////////////////
+/// return 0 for success
+/// return 1 for failure
+/// Check that the elements 
+/// are actually RooRealVar's
+/// If so, add them to the 
+/// list of vars
+Int_t ParamHistFunc::addVarSet( const RooArgList& vars ) {
+  for(auto const& comp : vars) {
+    if (!dynamic_cast<RooRealVar*>(comp)) {
+      coutE(InputArguments) << "ParamHistFunc::(" << GetName() << ") ERROR: component " 
+                            << comp->GetName() << " in variables list is not of type RooRealVar" 
+                            << std::endl;
+      RooErrorHandler::softAbort() ;
```

Comment:
Please remove that. The idea was that you would be dropped back to the terminal.

Better throw an exception, because that will actually drop you back to the terminal.

Reply:
Context: This used to work find in ROOT 5, but those times are gone.